src/esw/cameras.py

```diff
@@ -0,0 +1,351 @@
+        """
+        Adds endpoint to video source list and creates video_source if not existing.
+        :param endpoint: the endpoint (e.g. 10.0.0.7:5000)
+        :param args: a list of strings that are the arguments
+        """
+
+        assert not (endpoint in self.output_by_endpoint.keys())
+
+        if self.is_streaming():
+            # It exists and another stream is using it
+            # TODO - make sure that this check does what we want to do (self.resolution != args)
+            if self.resolution != args:
+                # If different args, just recreate video source and every output
+                try:
+                    # TODO - Make sure that the close does not break anything
+                    self.video_source.Close()
+                    self.video_source = jetson.utils.videoSource(f"/dev/video{self.device}", argv=args)
+                    self.video_source.Open()
+                    for other_endpoint in self.output_by_endpoint.values():
+                        self.output_by_endpoint[other_endpoint] = jetson.utils.videoOutput(
+                            f"rtp://{other_endpoint}", argv=args
+                        )
+                    self.output_by_endpoint[endpoint] = jetson.utils.videoOutput(f"rtp://{endpoint}", argv=args)
+                    self.resolution = args
+                except Exception:
+                    rospy.logerr(f"Failed to create video source for device {self.device}.")
+                    self.video_source = None
+                    self.output_by_endpoint = {}
+            else:
+                # If same args and endpoint is already being streamed to, then do nothing
+                if endpoint in self.output_by_endpoint.keys():
+                    return
+
+                # If same args, just create a new output
+                self.output_by_endpoint[endpoint] = jetson.utils.videoOutput(f"rtp://{endpoint}", argv=args)
+        else:
+            # If it does not exist already
+            try:
+                self.video_source = jetson.utils.videoSource(f"/dev/video{self.device}", argv=args)
+                self.video_source.Open()
+                self.output_by_endpoint[endpoint] = jetson.utils.videoOutput(f"rtp://{endpoint}", argv=args)
+                self.resolution = args
+            except Exception:
+                rospy.logerr(f"Failed to create video source for device {self.device}.")
+                self.video_source = None
+                self.output_by_endpoint = {}
+
+    def is_streaming(self) -> bool:
+        """
+        Returns if streaming or not
+        :return: a bool if streaming
+        """
+        is_streaming = len(self.output_by_endpoint) != 0
+        if is_streaming and self.video_source is not None:
+            return True
+        elif not is_streaming and self.video_source is None:
+            return False
+        else:
+            rospy.logerr("Should not have entered this state.")
```

Comment:
I'd be fine just asserting things like this. It shouldn't break on bad inputs to this program right? Just if internal state is mis-managed

Reply:
I do assert in the line right after